Editor: Prevent visibility from defaulting to 'public'

jeremeylduvall opened the pull request:
This is an anticipated fix for #3486, but I'm running into a little snag that I would love some advice on. The original issue was caused because we are using `visibility: 'public'` currently when we get the initial state, which means each post defaults to public regardless of the saved visibility. This PR uses `this.props.visibility` instead. Now, opening a post or refreshing a post maintains the visibility:

![editor](https://cloud.githubusercontent.com/assets/7240478/13375460/46707a0c-dd5d-11e5-830a-537d1a97a880.gif)

To test that piece:

1. Start a new post
2. Change the visibility to anything but public.
3. Refresh and/or exit/re-enter the editor for that post

Previously, you'll find that the post flipped to public. Now, it should maintain the set visibility. 

This created a second issue though that's been a bit of a bugger. When you use the 'Quick Add' icon in the masterbar to start a new post, the props are maintained so the new post opens up with the same visibility of the old post when I would expect it to be public by default. Here's a GIF of that happening:

![editor2](https://cloud.githubusercontent.com/assets/7240478/13375470/bd87dd6a-dd5d-11e5-9af4-f2b5f61b1a3a.gif)

That's a bigger issue to me than the original issue this intended to fix :)

I've tried:

1. Using `componentDidUpdate` to set the visibility back to 'public' when the editor loads for the new post. This did work, but I had to call `setState` inside of `componentDidUpdate`, which isn't the best and causes a lot of potential re-rendering.
2. I imported `isNew()` and used `PostEditStore.isNew()` to determine whether a post was new or not. If it was new, I set the visibility to 'public'. This worked as well, but it causes another issue when you first load a post. Since `isNew()` is true until the post is saved, the visibility constantly toggles back to 'public' before the autosave kicks in. An edge case most likely, but still something to consider.
3. Using a ternary operator in `getInitialState` to set visibility to 'public' if `isNew()` is true. The issue is that `getInitialState` isn't called because the page doesn't re-render when you use the 'Quick Add'.

Am I overlooking something simple?

blowery commented:
Could you do the work in `componentWillRecieveProps` instead of didUpdate?

nylen commented:
What is `this.props.visibility`?  How is it passed down to the `EditorVisibility` component, and how is it originally set in each of the following situations?

1. For a new post
2. For an existing post that is loaded into the editor

It looks to me like something is broken along the chain of properties passed in for an existing post.  The fix for this issue will involve finding the spot where the property is not being passed correctly, rather than changing the initial state.

jeremeylduvall commented:
Thank you both for the feedback. I have a much better understanding of the issue and a better solution. To recap, there are actually two issues that surfaced here:

**1. The correct visibility wasn't rendering when you opened a saved post.**

When the editor loads, we use `renderPostVisibility` in [`post-editor/editor-action-bar/index.jsx`](https://github.com/Automattic/wp-calypso/blob/40e6909c400428b9b1ce24b1ae20273d0c4a0afc/client/post-editor/editor-action-bar/index.jsx#L37) to grab the visibility for a post. That, in turn, uses `getVisibility` in [```lib/posts/utils.js```](https://github.com/Automattic/wp-calypso/blob/20167614982b4f44d95e12968cc1ecb42fddd1b5/client/lib/posts/utils.js#L137) for the work.

The issue when you open a new post is that we never actually make reference to `this.props.visibility` on the initial render so visibility is stuck to 'public' because of `getInitialState`. I added `componentWillMount`, which looks at whether `this.props.visibility !== this.state.visibility` (public by default) and updates if necessary.

**2. The visibility doesn't default back to public when you use the masterbar to start a new post.**

As an unintended consequence here, new posts initiated from the masterbar when you're already inside of a post render with the same visibility of the current post instead of defaulting to public. If you check the plumbing mentioned above though (`renderPostVisibility`), you'll notice that everything checks out for new posts (they're correctly interpreted as public). 

The underlying issue here (95% sure) is we're using `this.props.visibility` in `componentWillReceiveProps` which looks at the old value instead of the new value. So, when you load a new post, it grabs the old visibility value from the existing post and loads that value up. I changed that to look at `nextProps.visibility` instead. Now, new posts are rendering correctly because `nextProps.visibility` returns 'public'.

GIF of recent change:

![editor](https://cloud.githubusercontent.com/assets/7240478/13399484/c948bbec-dec0-11e5-808e-ed3e2107df35.gif)

jeremeylduvall commented:
Moved the work into a new function called `setVisibility` that's triggered on `componentWillMount` and on `componentWillReceiveProps`. The function performs the comparisons and then updates the state accordingly.

nylen commented:
Confirming this fixes #3486.  This component is still extremely buggy, but it's nothing new, so let's go ahead and get this fix in.  Reported what I'm seeing at #3673.
